crates/ntx-builder/src/state/mod.rs

```diff
@@ -170,7 +173,28 @@ impl State {
                 continue;
             }
 
-            let notes = account.notes().take(limit.get()).cloned().collect::<Vec<_>>();
+            // Remove notes that have failed too many times.
+            account.drain_failing_notes(Self::MAX_NOTE_ATTEMPTS);
```

Comment:
I would move this above the empty check since this might make the account empty.